### PR TITLE
feat: Implement minidump ingestion without symbolication

### PR DESCRIPTION
**Pull request by jan-auer**
This PR adds most basic support for Minidump crash reports just like [Mozilla Soccorro](https://wiki.mozilla.org/Socorro). This supports all clients using [Breakpad](https://chromium.googlesource.com/breakpad/breakpad/+/master/docs/getting_started_with_breakpad.md) and [Crashpad](https://chromium.googlesource.com/crashpad/crashpad/+/HEAD/doc/overview_design.md) libraries, including Electron. Symbolication is not supported yet, so frames will only show "<unknown>" and a memory address.

## Scope

So far, only event ingestion is implemented. This comprises:
 - A new "/api/*project*/minidump" endpoint to upload dumps and metadata
 - Extracting crash data, system infos and stacktraces from the dump into the event interface
 - A new "Minidump" model to store files for later downloading in the UI. It is uploaded just before pushing the event into the queue
 - Support the new "native" event type in the UI by rendering the iOS/Cocoa interface
 - Display the Minidump endpoint on the project keys settings page

## Notes

 - Crashpad (used in Electron) uploads minidumps with "Transfer-Encoding: Chunked". This PR includes a fix for supporting this, courtesy of @mitsuhiko 
 - Extracting event data from the minidump requires opening the file and running the processor on it. Without symbolication, this is a very fast process, so it is done synchronously. 
 - This PR requires symbolic 1.1.0 or later, which will be released any minute now
 - The event type is called `native` and not *minidump*, as this will become the common type for all these events. That will require some more refactoring and also design decisions, though.

## Next Steps

 - Extract CFI from debug symbols (cached!) and add another stackwalking pass to improve stacktraces. This should happen asynchronously in a preprocess step.
 - Support symbolication with Breakpad symbols. This will require some work in symbolic to generate symcaches from ASCII files.
 - Handle Breakpad symbols and split debug symbols (e.g. `.eh_frame` in MachO) in sentry-cli.

### CODE REVIEW

File: src/sentry/models/minidump.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+from sentry.constants import LOG_LEVELS_MAP
+from sentry.db.models import FlexibleForeignKey, Model, sane_repr
+from sentry.models.file import File
+
+
+class MinidumpFile(Model):
+    __core__ = False
+
+    file = FlexibleForeignKey('sentry.File')
+    event_id = models.CharField(max_length=36, unique=True)
```
Comment (mitsuhiko) [inline src/sentry/models/minidump.py]: This needs `index=True`.

Comment (mitsuhiko) [reply src/sentry/models/minidump.py]: Probably also a uniqueness constraint.

File: src/sentry/models/minidump.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                'instruction_addr': frame.instruction,
+                'function': '<unknown>',  # Required by interface
+            } for frame in thread.frames()],
+        },
+    } for thread in state.threads()]
+    data.setdefault('threads', {})['values'] = threads
+
+    # Mark the crashed thread and add its stacktrace to the exception
+    crashed_thread = threads[state.requesting_thread]
+    crashed_thread['crashed'] = True
```
Comment (mitsuhiko) [inline src/sentry/models/minidump.py]: the crashed thread should not have a stacktrace set. We need to remove it there. The stacktrace in that case is taken from the exception.

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR includes migrations</td>
    </tr>
  </tbody>
</table>

## Migration Checklist

- [ ] new columns need to be nullable (unless table is new)
- [ ] migration with any new index needs to be done concurrently
- [ ] data migrations should not be done inside a transaction
- [ ] before merging, check to make sure there aren't conflicting migration ids


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/web/api.py [hunk 3]
```diff
@@ -500,6 +502,63 @@ def process(self, request, project, key, auth, helper, data, **kwargs):
...
+        request.user = AnonymousUser()
+
+        project = self._get_project_from_id(project_id)
+        helper.context.bind_project(project)
+        Raven.tags_context(helper.context.get_tags_context())
+
+        # This is yanking the auth from the querystring since it's not
+        # in the POST body. This means we expect a `sentry_key` and
+        # `sentry_version` to be set in querystring
+        auth = helper.auth_from_request(request)
```
Comment (mitsuhiko) [inline src/sentry/web/api.py]: we should make this this work and add a test.

**Review by mitsuhiko — CHANGES_REQUESTED** [review]

File: src/sentry/web/api.py [hunk 4]
```diff
@@ -500,6 +502,63 @@ def process(self, request, project, key, auth, helper, data, **kwargs):
         return event_id
 
 
+class MinidumpView(StoreView):
+    helper_cls = MinidumpApiHelper
+    content_types = ('multipart/form-data')
```
Comment (mitsuhiko) [inline src/sentry/web/api.py]: Also this is wrong. Needs a trailing comma to make into a single item tuple.

**Review by mitsuhiko — CHANGES_REQUESTED** [review]

File: src/sentry/models/minidump.py (hunk 1 above)
Comment (jan-auer) [reply src/sentry/models/minidump.py]: `unique=True` should add a unique index, no?

Comment (mitsuhiko) [reply src/sentry/models/minidump.py]: yep

File: src/sentry/web/api.py (hunk 3 above)
Comment (jan-auer) [reply src/sentry/web/api.py]: Same goes for CSP below. Also, I noticed the comment (copy-pasted from CSP) is wrong, as `sentry_version` is not required in the querystring. It's not even checked anywhere. Would you like to see all this in this PR or a separate one?

Comment (mitsuhiko) [reply src/sentry/web/api.py]: You can add tests later. Just leave a note somewhere to not forget.

**Comment by jan-auer** [comment]
Should have all requested changes and updated dependency to symbolic.

**Review by mitsuhiko — APPROVED** [review]
Looks good. Please also bump the symbolic requirement in getsentry and enable the feature flag for superusers there after merging.

**Pull request merged**